docs/basic-component.md

```diff
@@ -0,0 +1,247 @@
+}
+```
+
+- Create and export an `interface` for your props in the format of `{PascalComponentName}Props`
+  - You can restrict a prop's type signature by redeclaring it (eg. changing a prop from `string | undefined` to `string`)
+  - If you need to change the type signature of an extended prop, ensure you `Omit` it otherwise it will become `never` (eg. changing a prop from `string | undefined` to `number`)
+- `Omit` `"className"` as we re-add it with an alias (`classNameOverride`)
+- Extend the native attributes of the closest HTML element of your component (eg. `<section>` will use `<div>` attributes)
+
+```tsx
+// Extending <section>
+import { HTMLAttributes } from 'react'
+export type SectionProps = Omit<HTMLAttributes<HTMLDivElement>, "className">
+
+// Extending <button>
+import { ButtonHTMLAttributes } from 'react'
+export type ButtonProps = Omit<ButtonHTMLAttributes<HTMLButtonElement>, "className">
+```
+
+- If your component is mainly extending another component, extend the props of that component and spread the props
+  - This will prevent the need for updating the consuming component with any new props in the child
+  - You don't need to extend HTML attributes as they would come from the child props (this also ensures consistency)
+
+```tsx
+export interface NewComponentProps extends PancakeStackProps {}
+
+export const NewComponent: React.VFC<NewComponentProps> = ({ ...props }) => <PancakeStack {...props} />
+```
+
+- Declare the `children` prop if you require it
+  - Usually it is `React.ReactNode`, however you can customise it to your needs
+- Add `classNameOverride` to replace the omitted `className`
+- Prefix boolean props with `is` or `has`
+- Declare a default value for optional boolean props
+  - Unless you specifically have a need to differentiate between `false` and `undefined`, this allows you to have the type safety of only needing to cater for a boolean value in any usages (eg. in a util)
+  - Aim to name the prop so that the default value is `false` which will allow the consumer to only need to include `isBoolean` as opposed to needing to negate it using `isBoolean={false}`
+- Prefix function props with `on`
+
+### The component
+
+```tsx
+import React from "react"
+
+export const PancakeStack: React.VFC<PancakeStackProps> = ({
+  children,
+  classNameOverride,
+  isBooleanProp,
+  hasOptionalBooleanProp = false,
+  onCustomFunction,
+  ...props
+}) => {
+  const [hasSyrup, setHasSyrup] = useState<boolean>(false)
+
+  const handleCustomFunction = () => {
+    onCustomFunction()
+    setHasSyrup(!hasSyrup)
+    return true
+  }
+
+  return (
+    <div
+      className={classnames(styles.pancakeStack, classNameOverride, {
+        [styles.someClass]: isBooleanProp,
+      })}
+      onSomething={handleCustomFunction}
+      {...props}
+    >
+      {children}
+      {doSomething(hasOptionalBooleanProp) && <SubComponent />}
+    </div>
+  );
+}
+```
+
+- Write and directly export a `React.VFC` (React VoidFunctionComponent)
```

Comment:
Probably include "if there is no children prop" for readability instead of having to click the link to find out.

Reply:
The point of `VFC` is that it will never include `children`, thus if you need it, you must always put it yourself. This increases type safety as `children` would not exist if not explicitly included (the flaw of `FC`).
I can put a summary of the difference between `VFC` and `FC` in here if that is clearer, but I had excluded it because I wasn't sure if it would be too much (since the docs are already kinda long).

Reply:
I think if we make it an onboarding item for all front end engineers to go through this then it should be a once off read and a refresher when things change. It could be a simple description of VFC if you don't have children FC if you do.

Reply:
Nope. Never `FC` is the idea :) If you ever need `children`, then you still use `VFC` and include `children` yourself in the props